libs/langchain/langchain/retrievers/document_compressors/encoded_chain_extract.py

```diff
@@ -0,0 +1,272 @@
+    return [para.strip() for para in paras if para.strip()]
+
+
+def get_default_tokenizer() -> Any:
+    return _make_spacy_pipeline_for_splitting(pipeline="en_core_web_sm")
+
+
+def number_sequences(text: str, len: int = 1, tokenizer: Any = None) -> str:
+    """
+    Number the sequences in a given text, preserving paragraph structure.
+    The default number of sentences per sequence is 1.
+
+    Args:
+        text: the text to number.
+        len: the number of sentences per sequence.
+
+    Returns:
+        the text with numbered sequences.
+    """
+    numbered_text = []
+    count = 0
+
+    if tokenizer is None:
+        tokenizer = get_default_tokenizer()
+    paragraphs = _split_paragraphs(text)
+    for paragraph in paragraphs:
+        sentences = [sent.text for sent in tokenizer(paragraph).sents]
+        for i, sentence in enumerate(sentences):
+            num = count // len + 1
+            number_prefix = f"#|{num}|#" if count % len == 0 else ""
+            sentence = f"{number_prefix} {sentence}"
+            count += 1
+            sentences[i] = sentence
+        numbered_paragraph = " ".join(sentences)
+        numbered_text.append(numbered_paragraph)
+
+    return "  \n\n  ".join(numbered_text)
+
+
+def default_get_input(query: str, context: str) -> Dict[str, str]:
+    """Return the compression chain input."""
+    return {"question": query, "context": context}
+
+
+def _get_default_chain_prompt() -> PromptTemplate:
+    template = prompt_template.format(no_output_str=NO_OUTPUT_STR)
+    return PromptTemplate(
+        template=template,
+        input_variables=["question", "context"],
+    )
+
+
+def extract_numbered_sequences(text: str, sequence_list: List[int]) -> str:
+    """
+    Extract specified sequences from a numbered text,
+    preserving paragraph structure.
+
+    Args:
+        text: the text to extract from.
+        sequence_list: the list of sequence numbers to extract.
+
+    Returns:
+        the extracted sequences from the text.
+    """
+
+    if not sequence_list:
+        return ""
+    sequence_pattern = re.compile(r"#\|(\d+)\|# ((?:(?!#\|).)+)")
+    paragraphs = _split_paragraphs(text)
+
+    extracted_paragraphs = []
+
+    for paragraph in paragraphs:
+        paragraph = paragraph.replace("\n", "\t")
+        sequences_with_numbers = sequence_pattern.findall(paragraph)
+
+        if extracted_sequences := [
+            sequence
+            for num, sequence in sequences_with_numbers
+            if int(num) in sequence_list
+        ]:
+            extracted_paragraphs.append(" ".join(extracted_sequences))
+
+    return "\n\n".join(extracted_paragraphs)
+
+
+class LLMEncodedChainExtractor(BaseDocumentCompressor):
+    """
+    Document compressor that uses an LLM chain to extract
+    the relevant parts of documents.
+    The relevant parts are encoded as sequence numbers
+    to optimize the extraction process.
+    """
+
+    llm_chain: LLMChain
```

Comment:
could we type this as Runnable instead and use `invoke` rather than `predict`? happy to help with that

Reply:
Have made this change based on my understanding of `Runnable`s. Take a look and let me know if you would do it differently.